contracts/utils/LowLevelCall.sol

```diff
@@ -0,0 +1,152 @@
+            result1 := mload(0)
+            result2 := mload(0x20)
+        }
+    }
+
+    /// === STATICCALL ===
+
+    /// @dev Performs a Solidity function call using a low level `staticcall` and ignoring the return data.
+    function staticcallRaw(address target, bytes memory data) internal view returns (bool success) {
+        assembly ("memory-safe") {
+            success := staticcall(gas(), target, add(data, 0x20), mload(data), 0, 0)
+        }
+    }
+
+    /// @dev Performs a Solidity function call using a low level `staticcall` and returns the first 32 bytes of the result
+    /// in the scratch space of memory. Useful for functions that return a single-word value.
+    ///
+    /// WARNING: Do not assume that the result is zero if `success` is false. Memory can be already allocated
+    /// and this function doesn't zero it out.
+    function staticcallReturnBytes32(
+        address target,
+        bytes memory data
+    ) internal view returns (bool success, bytes32 result) {
+        assembly ("memory-safe") {
+            success := staticcall(gas(), target, add(data, 0x20), mload(data), 0, 0x20)
+            result := mload(0)
+        }
+    }
+
+    /// @dev Performs a Solidity function call using a low level `staticcall` and returns the first 64 bytes of the result
+    /// in the scratch space of memory. Useful for functions that return a tuple of single-word values.
+    ///
+    /// WARNING: Do not assume that the results are zero if `success` is false. Memory can be already allocated
+    /// and this function doesn't zero it out.
+    function staticcallReturnBytes32Pair(
+        address target,
+        bytes memory data
+    ) internal view returns (bool success, bytes32 result1, bytes32 result2) {
+        assembly ("memory-safe") {
+            success := staticcall(gas(), target, add(data, 0x20), mload(data), 0, 0x40)
+            result1 := mload(0)
+            result2 := mload(0x20)
+        }
+    }
+
+    /// @dev Returns the size of the return data buffer.
+    function returnDataSize() internal pure returns (uint256 size) {
+        assembly ("memory-safe") {
+            size := returndatasize()
+        }
+    }
+
+    /// @dev Returns the return data buffer.
+    function returnData() internal pure returns (bytes memory data) {
+        uint256 size = returnDataSize();
+        assembly ("memory-safe") {
+            data := mload(0x40)
+            returndatacopy(data, 0, size)
+        }
+    }
+
+    /// === BUBBLE UP REVERT REASON ===
+
+    /**
+     * @dev Bubbles up the revert reason. This function is useful to bubble up the revert reason
+     * from a low level call. Developers can provide any bytes memory pointer to the revert reason
+     * since the function will revert regardless of the pointer value.
+     */
+    function bubbleRevert(bytes memory ptr) internal pure {
+        bubbleRevert(ptr.asPointer().asBytes32());
```

Comment:
if `bytes memory ptr` is not the same length as `returnDataSize`, this will not work.